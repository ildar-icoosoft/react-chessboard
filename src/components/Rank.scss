.rank {
  display: flex;
  flex-wrap: nowrap;
}
