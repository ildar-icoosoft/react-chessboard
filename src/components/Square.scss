.square {
  cursor: pointer;
  display: flex;
  justify-content: center;
}

.dark {
  background-color: rgb(181, 136, 99);

  &.isDragging {
    background-color: #646f40;
  }
}

.light {
  background-color: rgb(240, 217, 181);

  &.isDragging {
    background-color: #829769;
  }
}

.isDragging :global(.piece) {
  opacity: 0.3;
}
