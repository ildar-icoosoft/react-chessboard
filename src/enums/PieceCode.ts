export enum PieceCode {
  WHITE_PAWN = "wP",
  BLACK_PAWN = "bP",
  WHITE_KNIGHT = "wN",
  BLACK_KNIGHT = "bN",
  WHITE_BISHOP = "wB",
  BLACK_BISHOP = "bB",
  WHITE_ROOK = "wR",
  BLACK_ROOK = "bR",
  WHITE_QUEEN = "wQ",
  BLACK_QUEEN = "bQ",
  WHITE_KING = "wK",
  BLACK_KING = "bK",
}
