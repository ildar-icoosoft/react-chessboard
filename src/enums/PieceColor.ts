export enum PieceColor {
  WHITE = "white",
  BLACK = "black",
}
