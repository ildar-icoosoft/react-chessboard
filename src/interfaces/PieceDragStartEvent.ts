import { PieceCode } from "../enums/PieceCode";

export interface PieceDragStartEvent {
  coordinates: string;
  pieceCode: PieceCode;
}
