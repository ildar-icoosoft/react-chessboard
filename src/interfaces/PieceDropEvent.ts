import { PieceCode } from "../enums/PieceCode";

export interface PieceDropEvent {
  from: string;
  to: string;
  pieceCode: PieceCode;
  disableTransitionInNextPosition(): void;
}
