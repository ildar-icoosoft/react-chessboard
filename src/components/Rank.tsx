import React, { forwardRef, useImperativeHandle, useRef } from "react";
import { Square, SquareRef } from "./Square";
import { DEFAULT_BOARD_WIDTH, FILE_NAMES } from "../constants/constants";
import { PieceColor } from "../enums/PieceColor";
import { Position } from "../interfaces/Position";
import { SquareCssClasses } from "../interfaces/SquareCssClasses";
import { PieceCode } from "../enums/PieceCode";
import { PieceDropEvent } from "../interfaces/PieceDropEvent";
import css from "./Rank.scss";
import { PieceDragStartEvent } from "../interfaces/PieceDragStartEvent";
import { XYCoordinates } from "../interfaces/XYCoordinates";
import { SquareTransitionFrom } from "../interfaces/SquareTransitionFrom";

export interface RankRef {
  getSquareXYCoordinates(coordinates: string): XYCoordinates;
}

export interface RankProps {
  rankName: string;
  position?: Position;
  orientation?: PieceColor;
  squareCssClasses?: SquareCssClasses;
  dragStartCssClass?: string[] | string;
  dragEnterSquareCssClass?: string[] | string;
  draggable?: boolean;
  width?: number;
  showNotation?: boolean;
  transitionDuration?: number;
  transitionPieces?: Record<string, SquareTransitionFrom>;
  allowDrag?: (pieceCode: PieceCode, coordinates: string) => boolean;

  onSquareClick?(coordinates: string): void;

  onSquareRightClick?(coordinates: string): void;

  onDragStart?(event: PieceDragStartEvent): void;

  onDragEnterSquare?(coordinates: string): void;

  onDrop?(event: PieceDropEvent): void;

  onMouseEnterSquare?(coordinates: string): void;

  onMouseLeaveSquare?(coordinates: string): void;
}

export const Rank = forwardRef<RankRef, RankProps>(
  (
    {
      draggable = false,
      width = DEFAULT_BOARD_WIDTH,
      showNotation = true,
      rankName,
      position = {},
      squareCssClasses = {},
      dragStartCssClass,
      dragEnterSquareCssClass,
      orientation = PieceColor.WHITE,
      transitionDuration,
      transitionPieces = {},
      allowDrag,
      onSquareClick,
      onSquareRightClick,
      onDragStart,
      onDragEnterSquare,
      onDrop,
      onMouseEnterSquare,
      onMouseLeaveSquare,
    },
    ref
  ) => {
    const squareRefs = useRef<Record<string, SquareRef | null>>({});

    useImperativeHandle(ref, () => ({
      getSquareXYCoordinates(coordinates: string): XYCoordinates {
        if (!squareRefs.current[coordinates]) {
          throw Error(`Square ${coordinates} not found`);
        }

        return (squareRefs.current[
          coordinates
        ] as SquareRef).getXYCoordinates();
      },
    }));

    let fileNames: string[];
    if (orientation === PieceColor.WHITE) {
      fileNames = FILE_NAMES;
    } else {
      fileNames = FILE_NAMES.slice().reverse();
    }

    return (
      <div
        data-testid={`rank-${rankName}`}
        className={css.rank}
        style={{ width: `${width}px` }}
      >
        {fileNames.map((fileName) => {
          const coordinates: string = fileName + rankName;

          return (
            <Square
              ref={(el) => (squareRefs.current[coordinates] = el)}
              allowDrag={allowDrag}
              coordinates={coordinates}
              transitionDuration={transitionDuration}
              transitionFrom={transitionPieces[coordinates]}
              draggable={draggable}
              width={width / 8}
              showNotation={showNotation}
              key={fileName}
              pieceCode={position[coordinates]}
              cssClass={squareCssClasses[coordinates]}
              dragStartCssClass={dragStartCssClass}
              dragEnterSquareCssClass={dragEnterSquareCssClass}
              orientation={orientation}
              onSquareClick={onSquareClick}
              onSquareRightClick={onSquareRightClick}
              onDragStart={onDragStart}
              onDragEnterSquare={onDragEnterSquare}
              onDrop={onDrop}
              onMouseEnterSquare={onMouseEnterSquare}
              onMouseLeaveSquare={onMouseLeaveSquare}
            />
          );
        })}
      </div>
    );
  }
);
