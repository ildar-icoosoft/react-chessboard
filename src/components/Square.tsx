import React, {
  forwardRef,
  MouseEvent,
  useEffect,
  useImperativeHandle,
  useRef,
} from "react";
import { Piece } from "./Piece";
import classNames from "classnames";
import css from "./Square.scss";
import { isLightSquare } from "../utils/chess";
import { PieceCode } from "../enums/PieceCode";
import { useDrag, useDrop } from "react-dnd";
import { DragItemType } from "../enums/DragItemType";
import { getEmptyImage } from "react-dnd-html5-backend";
import { useCombinedRefs } from "../hooks/useCombinedRefs";
import { PieceDragObjectOld } from "../interfaces/PieceDragObjectOld";
import { PieceDropEvent } from "../interfaces/PieceDropEvent";
import { PieceDragStartEvent } from "../interfaces/PieceDragStartEvent";
import { Notation } from "./Notation";
import { PieceColor } from "../enums/PieceColor";
import { Identifier } from "dnd-core";
import { XYCoordinates } from "../interfaces/XYCoordinates";
import { SquareTransitionFrom } from "../interfaces/SquareTransitionFrom";
import { PhantomPiece } from "./PhantomPiece";
import { DEFAULT_SQUARE_WIDTH } from "../constants/constants";

export interface SquareRef {
  getDragHandlerId(): Identifier | null;
  getDropHandlerId(): Identifier | null;
  getXYCoordinates(): XYCoordinates;
}

export interface SquareProps {
  allowDrag?: (pieceCode: PieceCode, coordinates: string) => boolean;
  coordinates: string;
  pieceCode?: PieceCode;
  cssClass?: string[] | string;
  transitionFrom?: SquareTransitionFrom;
  dragStartCssClass?: string[] | string;
  dragEnterSquareCssClass?: string[] | string;
  draggable?: boolean;
  width?: number;
  transitionDuration?: number;
  showNotation?: boolean;
  orientation?: PieceColor;

  onSquareClick?(coordinates: string): void;

  onSquareRightClick?(coordinates: string): void;

  onDragStart?(event: PieceDragStartEvent): void;

  onDragEnterSquare?(coordinates: string): void;

  onDrop?(event: PieceDropEvent): void;

  onMouseEnterSquare?(coordinates: string): void;

  onMouseLeaveSquare?(coordinates: string): void;
}

export const Square = forwardRef<SquareRef, SquareProps>(
  (
    {
      allowDrag,
      coordinates,
      transitionDuration,
      draggable = false,
      width = DEFAULT_SQUARE_WIDTH,
      showNotation = true,
      pieceCode,
      cssClass = [],
      dragStartCssClass = [],
      dragEnterSquareCssClass = [],
      orientation = PieceColor.WHITE,
      onSquareClick,
      onSquareRightClick,
      onDragStart,
      onDragEnterSquare,
      onDrop,
      onMouseEnterSquare,
      onMouseLeaveSquare,
      transitionFrom,
    },
    ref
  ) => {
    const [{ isOver, dropHandlerId }, dropRef] = useDrop({
      accept: DragItemType.PIECE,
      drop(item: PieceDragObjectOld) {
        if (onDrop) {
          onDrop({
            sourceCoordinates: item.coordinates,
            targetCoordinates: coordinates,
            pieceCode: item.pieceCode,
          });
        }
      },
      hover(_item: PieceDragObjectOld, monitor) {
        // we could also use HTML onDragEnter event, but for some reason it called twice on beginning of the drag
        if (!monitor.isOver({ shallow: true })) {
          if (onDragEnterSquare) {
            onDragEnterSquare(coordinates);
          }
        }
      },
      collect(monitor) {
        return {
          isOver: monitor.isOver(),
          dropHandlerId: monitor.getHandlerId(),
        };
      },
    });

    const handleSquareClick = (): void => {
      if (onSquareClick) {
        onSquareClick(coordinates);
      }
    };

    const handleContextMenu = (
      event: MouseEvent<HTMLTableDataCellElement>
    ): void => {
      event.preventDefault();

      if (onSquareRightClick) {
        onSquareRightClick(coordinates);
      }
    };

    const handleMouseEnterSquare = (): void => {
      if (onMouseEnterSquare) {
        onMouseEnterSquare(coordinates);
      }
    };

    const handleMouseLeaveSquare = (): void => {
      if (onMouseLeaveSquare) {
        onMouseLeaveSquare(coordinates);
      }
    };

    const isLight = isLightSquare(coordinates);

    const [{ isDragging, dragHandlerId }, dragRef, preview] = useDrag({
      begin(_monitor) {
        if (onDragStart) {
          onDragStart({
            coordinates,
            pieceCode: pieceCode as PieceCode,
          });
        }
      },
      item: {
        type: DragItemType.PIECE,
        pieceCode,
        coordinates,
      },
      collect: (monitor) => ({
        isDragging: monitor.isDragging(),
        dragHandlerId: monitor.getHandlerId(),
      }),
      canDrag() {
        if (!pieceCode) {
          return false;
        }
        if (allowDrag) {
          return allowDrag(pieceCode, coordinates);
        }
        return draggable;
      },
    });

    const domRef = useRef<HTMLElement>();

    useImperativeHandle(ref, () => ({
      getDragHandlerId: (): Identifier | null => dragHandlerId,
      getDropHandlerId: (): Identifier | null => dropHandlerId,
      getXYCoordinates(): XYCoordinates {
        const el: HTMLElement = domRef.current as HTMLElement;

        // for unit testing where getBoundingClientRect is undefined
        if (!el || !el.getBoundingClientRect) {
          return {
            x: 0,
            y: 0,
          };
        }

        const { x, y } = el.getBoundingClientRect();
        return {
          x: x || 0,
          y: y || 0,
        };
      },
    }));

    useEffect(() => {
      preview(getEmptyImage());
    });

    return (
      <div
        data-testid={`square-${coordinates}`}
        ref={useCombinedRefs(dropRef, dragRef, domRef)}
        style={{ width: `${width}px`, height: `${width}px` }}
        className={classNames(
          css.square,
          cssClass,
          isOver && dragEnterSquareCssClass,
          isDragging && dragStartCssClass,
          {
            [css.light]: isLight,
            [css.dark]: !isLight,
            [css.isDragging]: isDragging,
            [css.isOver]: isOver,
          }
        )}
        onClick={handleSquareClick}
        onContextMenu={handleContextMenu}
        onMouseEnter={handleMouseEnterSquare}
        onMouseLeave={handleMouseLeaveSquare}
      >
        {pieceCode && (
          <Piece
            pieceCode={pieceCode}
            transitionDuration={transitionDuration}
            width={width}
            transitionFrom={transitionFrom}
            key={
              transitionFrom ? `piece-${transitionFrom.algebraic}` : undefined
            }
          />
        )}
        {transitionFrom && transitionFrom.phantomPiece && (
          <PhantomPiece
            pieceCode={transitionFrom.phantomPiece}
            transitionDuration={transitionDuration}
            width={width}
            xYCoordinates={{ x: 0, y: 0 }}
            key={`phantom-piece-${transitionFrom.algebraic}`}
          />
        )}
        {showNotation && (
          <Notation
            coordinates={coordinates}
            orientation={orientation}
            width={width}
          />
        )}
      </div>
    );
  }
);
