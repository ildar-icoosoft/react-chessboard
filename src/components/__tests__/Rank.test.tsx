import React, { createRef } from "react";
import TestRenderer from "react-test-renderer";
import { Rank, RankRef } from "../Rank";
import { Square } from "../Square";
import { PieceCode } from "../../enums/PieceCode";
import { PieceColor } from "../../enums/PieceColor";
import { Position } from "../../interfaces/Position";
import { SquareCssClasses } from "../../interfaces/SquareCssClasses";
import { PieceDropEvent } from "../../interfaces/PieceDropEvent";
import { PieceDragStartEvent } from "../../interfaces/PieceDragStartEvent";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { wrapInTestContext } from "react-dnd-test-utils";
import { ReactDndRefType } from "../../interfaces/ReactDndRefType";

jest.useFakeTimers();

describe("Rank", () => {
  const RankWithDnd = wrapInTestContext(Rank);

  it("Snapshot", () => {
    const tree = TestRenderer.create(<RankWithDnd rankName={"2"} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("children components", () => {
    it("contains 8 Squares", () => {
      const testInstance = TestRenderer.create(<RankWithDnd rankName={"2"} />)
        .root;

      expect(testInstance.findAllByType(Square).length).toBe(8);
    });
  });

  describe("children components props", () => {
    describe("Square", () => {
      it("coordinates for white orientation", () => {
        const testInstance = TestRenderer.create(
          <RankWithDnd rankName={"2"} orientation={PieceColor.WHITE} />
        ).root;

        const squares: TestRenderer.ReactTestInstance[] = testInstance.findAllByType(
          Square
        );

        const files: string[] = ["a", "b", "c", "d", "e", "f", "g", "h"];
        for (let i = 0; i < 8; i++) {
          expect(squares[i].props.coordinates).toBe(files[i] + "2");
        }
      });

      it("coordinates for black orientation", () => {
        const testInstance = TestRenderer.create(
          <RankWithDnd rankName={"2"} orientation={PieceColor.BLACK} />
        ).root;

        const squares: TestRenderer.ReactTestInstance[] = testInstance.findAllByType(
          Square
        );

        const files: string[] = ["a", "b", "c", "d", "e", "f", "g", "h"];
        for (let i = 0; i < 8; i++) {
          expect(squares[i].props.coordinates).toBe(files[7 - i] + "2");
        }
      });

      it("pieceCode", () => {
        const position: Position = {
          a1: PieceCode.BLACK_KING,
          b2: PieceCode.WHITE_QUEEN,
          c3: PieceCode.WHITE_KNIGHT,
        };

        const testInstance = TestRenderer.create(
          <RankWithDnd rankName={"2"} position={position} />
        ).root;

        const square1: TestRenderer.ReactTestInstance = testInstance.findByProps(
          {
            coordinates: "b2",
          }
        );
        expect(square1.props.pieceCode).toBe(PieceCode.WHITE_QUEEN);

        const square2: TestRenderer.ReactTestInstance = testInstance.findByProps(
          {
            coordinates: "g2",
          }
        );
        expect(square2.props.pieceCode).toBeUndefined();
      });

      it("squareCssClasses", () => {
        const squareCssClasses: SquareCssClasses = {
          e2: ["class1", "class2"],
        };

        const testInstance = TestRenderer.create(
          <RankWithDnd rankName={"2"} squareCssClasses={squareCssClasses} />
        ).root;

        const square: TestRenderer.ReactTestInstance = testInstance.findByProps(
          {
            coordinates: "e2",
          }
        );

        expect(square.props.cssClass).toEqual(["class1", "class2"]);

        const square2: TestRenderer.ReactTestInstance = testInstance.findByProps(
          {
            coordinates: "h2",
          }
        );

        expect(square2.props.cssClass).toBeUndefined();
      });

      it("dragStartCssClass", () => {
        const dragStartCssClass: string[] = ["class1", "class2"];

        const testInstance = TestRenderer.create(
          <RankWithDnd rankName={"2"} dragStartCssClass={dragStartCssClass} />
        ).root;

        const square: TestRenderer.ReactTestInstance = testInstance.findByProps(
          {
            coordinates: "e2",
          }
        );

        expect(square.props.dragStartCssClass).toEqual(["class1", "class2"]);
      });

      it("dragEnterSquareCssClass", () => {
        const dragEnterSquareCssClass: string[] = ["class1", "class2"];

        const testInstance = TestRenderer.create(
          <RankWithDnd
            rankName={"2"}
            dragEnterSquareCssClass={dragEnterSquareCssClass}
          />
        ).root;

        const square: TestRenderer.ReactTestInstance = testInstance.findByProps(
          {
            coordinates: "e2",
          }
        );

        expect(square.props.dragEnterSquareCssClass).toEqual([
          "class1",
          "class2",
        ]);
      });

      it("transitionDuration", () => {
        const testRenderer = TestRenderer.create(
          <RankWithDnd rankName={"2"} />
        );
        const testInstance = testRenderer.root;

        const square: TestRenderer.ReactTestInstance = testInstance.findByProps(
          {
            coordinates: "e2",
          }
        );

        expect(square.props.transitionDuration).toBeUndefined();

        testRenderer.update(
          <RankWithDnd rankName={"2"} transitionDuration={600} />
        );

        expect(square.props.transitionDuration).toBe(600);
      });

      it("draggable", () => {
        const testRenderer = TestRenderer.create(
          <RankWithDnd rankName={"2"} draggable={true} />
        );
        const testInstance = testRenderer.root;

        const square: TestRenderer.ReactTestInstance = testInstance.findByProps(
          {
            coordinates: "e2",
          }
        );

        expect(square.props.draggable).toBeTruthy();

        testRenderer.update(<RankWithDnd rankName={"2"} />);

        expect(square.props.draggable).toBeFalsy();
      });

      it("allowDrag", () => {
        const allowDrag = jest.fn(() => true);

        const testRenderer = TestRenderer.create(
          <RankWithDnd rankName={"2"} allowDrag={allowDrag} />
        );
        const testInstance = testRenderer.root;

        const square: TestRenderer.ReactTestInstance = testInstance.findByProps(
          {
            coordinates: "e2",
          }
        );

        expect(square.props.allowDrag).toBe(allowDrag);

        testRenderer.update(<RankWithDnd rankName={"2"} />);

        expect(square.props.allowDrag).toBeUndefined();
      });

      it("showNotation", () => {
        const testRenderer = TestRenderer.create(
          <RankWithDnd rankName={"2"} />
        );
        const testInstance = testRenderer.root;

        const square: TestRenderer.ReactTestInstance = testInstance.findByProps(
          {
            coordinates: "e2",
          }
        );

        expect(square.props.showNotation).toBeTruthy();

        testRenderer.update(
          <RankWithDnd rankName={"2"} showNotation={false} />
        );

        expect(square.props.showNotation).toBeFalsy();
      });

      it("orientation", () => {
        const testRenderer = TestRenderer.create(
          <RankWithDnd rankName={"2"} />
        );
        const testInstance = testRenderer.root;

        const square: TestRenderer.ReactTestInstance = testInstance.findByProps(
          {
            coordinates: "b2",
          }
        );
        expect(square.props.orientation).toBe(PieceColor.WHITE);

        testRenderer.update(
          <RankWithDnd rankName={"2"} orientation={PieceColor.BLACK} />
        );

        expect(square.props.orientation).toBe(PieceColor.BLACK);
      });

      it("width", () => {
        const testRenderer = TestRenderer.create(
          <RankWithDnd rankName={"2"} />
        );
        const testInstance = testRenderer.root;

        const square: TestRenderer.ReactTestInstance = testInstance.findByProps(
          {
            coordinates: "b2",
          }
        );
        expect(square.props.width).toBe(480 / 8);

        testRenderer.update(<RankWithDnd rankName={"2"} width={240} />);

        expect(square.props.width).toBe(240 / 8);
      });

      it("transitionFrom", () => {
        const testRenderer = TestRenderer.create(
          <RankWithDnd rankName={"4"} />
        );
        const testInstance = testRenderer.root;

        const square: TestRenderer.ReactTestInstance = testInstance.findByProps(
          {
            coordinates: "e4",
          }
        );
        expect(square.props.transitionFrom).toBeUndefined();

        testRenderer.update(
          <RankWithDnd
            rankName={"4"}
            transitionPieces={{
              e4: {
                algebraic: "e2",
                x: 0,
                y: 0,
              },
            }}
          />
        );

        expect(square.props.transitionFrom).toEqual({
          algebraic: "e2",
          x: 0,
          y: 0,
        });
      });
    });
  });

  describe("Events", () => {
    it("onSquareClick", () => {
      const onSquareClick = jest.fn();

      const testInstance = TestRenderer.create(
        <RankWithDnd rankName={"2"} onSquareClick={onSquareClick} />
      ).root;

      const square: TestRenderer.ReactTestInstance = testInstance.findByProps({
        coordinates: "e2",
      });

      TestRenderer.act(() => {
        square.props.onSquareClick("e2");
      });

      expect(onSquareClick).toHaveBeenCalledTimes(1);

      expect(onSquareClick).toBeCalledWith("e2");
    });

    it("onSquareRightClick", () => {
      const onSquareRightClick = jest.fn();

      const testInstance = TestRenderer.create(
        <RankWithDnd rankName={"2"} onSquareRightClick={onSquareRightClick} />
      ).root;

      const square: TestRenderer.ReactTestInstance = testInstance.findByProps({
        coordinates: "e2",
      });

      TestRenderer.act(() => {
        square.props.onSquareRightClick("e2");
      });

      expect(onSquareRightClick).toHaveBeenCalledTimes(1);

      expect(onSquareRightClick).toBeCalledWith("e2");
    });

    it("onDragStart", () => {
      const onDragStart = jest.fn();

      const testInstance = TestRenderer.create(
        <RankWithDnd rankName={"2"} onDragStart={onDragStart} />
      ).root;

      const square: TestRenderer.ReactTestInstance = testInstance.findByProps({
        coordinates: "e2",
      });

      const dragStartEvent: PieceDragStartEvent = {
        coordinates: "e2",
        pieceCode: PieceCode.WHITE_PAWN,
      };

      TestRenderer.act(() => {
        square.props.onDragStart(dragStartEvent);
      });

      expect(onDragStart).toHaveBeenCalledTimes(1);

      expect(onDragStart).toBeCalledWith(dragStartEvent);
    });

    it("onDragEnterSquare", () => {
      const onDragEnterSquare = jest.fn();

      const testInstance = TestRenderer.create(
        <RankWithDnd rankName={"2"} onDragEnterSquare={onDragEnterSquare} />
      ).root;

      const square: TestRenderer.ReactTestInstance = testInstance.findByProps({
        coordinates: "e2",
      });

      TestRenderer.act(() => {
        square.props.onDragEnterSquare("e2");
      });

      expect(onDragEnterSquare).toHaveBeenCalledTimes(1);

      expect(onDragEnterSquare).toBeCalledWith("e2");
    });

    it("onDrop", () => {
      const onDrop = jest.fn();

      const testInstance = TestRenderer.create(
        <RankWithDnd rankName={"2"} onDrop={onDrop} />
      ).root;

      const square: TestRenderer.ReactTestInstance = testInstance.findByProps({
        coordinates: "e2",
      });

      const dropEvent: PieceDropEvent = {
        sourceCoordinates: "e2",
        targetCoordinates: "e4",
        pieceCode: PieceCode.WHITE_PAWN,
      };

      TestRenderer.act(() => {
        square.props.onDrop(dropEvent);
      });

      expect(onDrop).toHaveBeenCalledTimes(1);

      expect(onDrop).toBeCalledWith(dropEvent);
    });

    it("onMouseEnterSquare", () => {
      const onMouseEnterSquare = jest.fn();

      const testInstance = TestRenderer.create(
        <RankWithDnd rankName={"2"} onMouseEnterSquare={onMouseEnterSquare} />
      ).root;

      const square: TestRenderer.ReactTestInstance = testInstance.findByProps({
        coordinates: "e2",
      });

      TestRenderer.act(() => {
        square.props.onMouseEnterSquare("e2");
      });

      expect(onMouseEnterSquare).toHaveBeenCalledTimes(1);

      expect(onMouseEnterSquare).toBeCalledWith("e2");
    });

    it("onMouseLeaveSquare", () => {
      const onMouseLeaveSquare = jest.fn();

      const testInstance = TestRenderer.create(
        <RankWithDnd rankName={"2"} onMouseLeaveSquare={onMouseLeaveSquare} />
      ).root;

      const square: TestRenderer.ReactTestInstance = testInstance.findByProps({
        coordinates: "e2",
      });

      TestRenderer.act(() => {
        square.props.onMouseLeaveSquare("e2");
      });

      expect(onMouseLeaveSquare).toHaveBeenCalledTimes(1);

      expect(onMouseLeaveSquare).toBeCalledWith("e2");
    });
  });

  describe("methods", () => {
    it("getSquareXYCoordinates()", () => {
      const dragAndDropRef = createRef<ReactDndRefType>();

      render(<RankWithDnd rankName={"2"} ref={dragAndDropRef} />);

      const rankRef: RankRef = (dragAndDropRef.current as ReactDndRefType).getDecoratedComponent<
        RankRef
      >();

      expect(rankRef.getSquareXYCoordinates("b2")).toEqual({
        x: 0,
        y: 0,
      });
    });

    it("getSquareXYCoordinates() throws Error", () => {
      const dragAndDropRef = createRef<ReactDndRefType>();

      render(<RankWithDnd rankName={"2"} ref={dragAndDropRef} />);

      const rankRef: RankRef = (dragAndDropRef.current as ReactDndRefType).getDecoratedComponent<
        RankRef
      >();

      expect(() => rankRef.getSquareXYCoordinates("c3")).toThrow();
    });
  });

  describe("DOM structure", () => {
    it("must contain a div", () => {
      const { queryByTestId } = render(<RankWithDnd rankName={"2"} />);
      expect(queryByTestId("rank-2")).toBeInTheDocument();
    });
  });
});
