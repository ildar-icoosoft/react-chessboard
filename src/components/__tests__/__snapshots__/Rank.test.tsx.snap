// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rank Snapshot 1`] = `
<div
  className="rank"
  data-testid="rank-2"
  style={
    Object {
      "width": "480px",
    }
  }
>
  <div
    className="square light"
    data-testid="square-a2"
    onClick={[Function]}
    onContextMenu={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "height": "60px",
        "width": "60px",
      }
    }
  >
    <div
      className="rankNotation lightSquare"
      data-testid="notation-rank-2"
      style={
        Object {
          "fontSize": "10px",
          "paddingRight": "50px",
        }
      }
    >
      2
    </div>
  </div>
  <div
    className="square dark"
    data-testid="square-b2"
    onClick={[Function]}
    onContextMenu={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "height": "60px",
        "width": "60px",
      }
    }
  />
  <div
    className="square light"
    data-testid="square-c2"
    onClick={[Function]}
    onContextMenu={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "height": "60px",
        "width": "60px",
      }
    }
  />
  <div
    className="square dark"
    data-testid="square-d2"
    onClick={[Function]}
    onContextMenu={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "height": "60px",
        "width": "60px",
      }
    }
  />
  <div
    className="square light"
    data-testid="square-e2"
    onClick={[Function]}
    onContextMenu={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "height": "60px",
        "width": "60px",
      }
    }
  />
  <div
    className="square dark"
    data-testid="square-f2"
    onClick={[Function]}
    onContextMenu={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "height": "60px",
        "width": "60px",
      }
    }
  />
  <div
    className="square light"
    data-testid="square-g2"
    onClick={[Function]}
    onContextMenu={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "height": "60px",
        "width": "60px",
      }
    }
  />
  <div
    className="square dark"
    data-testid="square-h2"
    onClick={[Function]}
    onContextMenu={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "height": "60px",
        "width": "60px",
      }
    }
  />
</div>
`;
